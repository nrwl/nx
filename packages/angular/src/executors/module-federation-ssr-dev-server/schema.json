{
  "$schema": "http://json-schema.org/draft-07/schema",
  "title": "Module Federation SSR Dev Server Target",
  "continuous": true,
  "outputCapture": "direct-nodejs",
  "description": "The module-federation-ssr-dev-server executor is reserved exclusively for use with host SSR Module Federation applications. It allows the user to specify which remote applications should be served with the host.",
  "type": "object",
  "properties": {
    "browserTarget": {
      "type": "string",
      "description": "Browser target to build.",
      "pattern": ".+:.+(:.+)?"
    },
    "serverTarget": {
      "type": "string",
      "description": "Server target to build.",
      "pattern": ".+:.+(:.+)?"
    },
    "host": {
      "type": "string",
      "description": "Host to listen on.",
      "default": "localhost"
    },
    "port": {
      "type": "number",
      "default": 4200,
      "description": "Port to start the development server at. Default is 4200. Pass 0 to get a dynamically assigned port."
    },
    "publicHost": {
      "type": "string",
      "description": "The URL that the browser client should use to connect to the development server. Use for a complex dev server setup, such as one with reverse proxies."
    },
    "open": {
      "type": "boolean",
      "description": "Opens the url in default browser.",
      "default": false,
      "alias": "o"
    },
    "progress": {
      "type": "boolean",
      "description": "Log progress to the console while building."
    },
    "inspect": {
      "type": "boolean",
      "description": "Launch the development server in inspector mode and listen on address and port '127.0.0.1:9229'.",
      "default": false
    },
    "ssl": {
      "type": "boolean",
      "description": "Serve using HTTPS.",
      "default": false
    },
    "sslKey": {
      "type": "string",
      "description": "SSL key to use for serving HTTPS."
    },
    "sslCert": {
      "type": "string",
      "description": "SSL certificate to use for serving HTTPS."
    },
    "proxyConfig": {
      "type": "string",
      "description": "Proxy configuration file."
    },
    "devRemotes": {
      "type": "array",
      "items": {
        "oneOf": [
          {
            "type": "string"
          },
          {
            "type": "object",
            "properties": {
              "remoteName": {
                "type": "string"
              },
              "configuration": {
                "type": "string"
              }
            },
            "required": ["remoteName"],
            "additionalProperties": false
          }
        ]
      },
      "description": "List of remote applications to run in development mode (i.e. using serve target).",
      "x-priority": "important"
    },
    "skipRemotes": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "description": "List of remote applications to not automatically serve, either statically or in development mode."
    },
    "verbose": {
      "type": "boolean",
      "description": "Adds more details to output logging.",
      "default": false
    },
    "pathToManifestFile": {
      "type": "string",
      "description": "Path to a Module Federation manifest file (e.g. `my/path/to/module-federation.manifest.json`) containing the dynamic remote applications relative to the workspace root."
    },
    "isInitialHost": {
      "type": "boolean",
      "description": "Whether the host that is running this executor is the first in the project tree to do so.",
      "default": true,
      "x-priority": "internal"
    },
    "parallel": {
      "type": "number",
      "description": "Max number of parallel processes for building static remotes"
    },
    "staticRemotesPort": {
      "type": "number",
      "description": "The port at which to serve the file-server for the static remotes."
    }
  },
  "additionalProperties": false,
  "required": ["browserTarget", "serverTarget"]
}
