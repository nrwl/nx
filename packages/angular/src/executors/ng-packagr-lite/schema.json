{
  "version": 2,
  "outputCapture": "direct-nodejs",
  "$schema": "https://json-schema.org/schema",
  "title": "ng-packagr Target",
  "description": "Builds an Angular library with support for incremental builds.\n\nThis executor is meant to be used with buildable libraries in an incremental build scenario. It is similar to the `@nx/angular:package` executor but it only produces ESM2022 bundles.",
  "cli": "nx",
  "type": "object",
  "presets": [
    {
      "name": "Updating Project Dependencies for Buildable Library",
      "keys": ["project"]
    }
  ],
  "properties": {
    "project": {
      "type": "string",
      "description": "The file path for the ng-packagr configuration file, relative to the workspace root."
    },
    "tsConfig": {
      "type": "string",
      "description": "The full path for the TypeScript configuration file, relative to the workspace root.",
      "x-completion-type": "file",
      "x-completion-glob": "tsconfig.*.json",
      "x-priority": "important"
    },
    "watch": {
      "type": "boolean",
      "description": "Whether to run a build when any file changes.",
      "default": false
    },
    "poll": {
      "type": "number",
      "description": "Enable and define the file watching poll time period in milliseconds. _Note: this is only supported in Angular versions >= 18.0.0_."
    }
  },
  "additionalProperties": false,
  "required": ["project"]
}
