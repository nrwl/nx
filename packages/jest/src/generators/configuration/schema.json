{
  "$schema": "https://json-schema.org/schema",
  "$id": "NxJestProject",
  "cli": "nx",
  "title": "Add Jest Configuration to a project",
  "description": "Add Jest Configuration to a project.",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      },
      "x-priority": "important"
    },
    "skipSetupFile": {
      "type": "boolean",
      "description": "Skips the setup file required for angular.",
      "default": false,
      "x-deprecated": "Use `--setup-file` instead."
    },
    "setupFile": {
      "type": "string",
      "enum": ["none", "angular", "web-components"],
      "description": "The setup file to be generated.",
      "default": "none",
      "x-priority": "important"
    },
    "skipSerializers": {
      "type": "boolean",
      "description": "Skips the serializers required to snapshot angular templates.",
      "default": false
    },
    "supportTsx": {
      "type": "boolean",
      "description": "Setup `tsx` support.",
      "default": false,
      "x-priority": "important"
    },
    "testEnvironment": {
      "type": "string",
      "enum": ["jsdom", "node", "none"],
      "description": "The test environment for jest.",
      "default": "jsdom",
      "x-priority": "important"
    },
    "compiler": {
      "type": "string",
      "enum": ["tsc", "babel", "swc"],
      "description": "The compiler to use for source and tests.",
      "default": "tsc"
    },
    "babelJest": {
      "type": "boolean",
      "alias": "babel-jest",
      "description": "Use `babel-jest` instead of `ts-jest`.",
      "x-deprecated": "Use `--compiler=babel` instead.",
      "default": false
    },
    "skipFormat": {
      "description": "Skip formatting files",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description": "Do not add dependencies to `package.json`.",
      "x-priority": "internal"
    },
    "js": {
      "type": "boolean",
      "default": false,
      "description": "Use JavaScript instead of TypeScript for config files"
    },
    "runtimeTsconfigFileName": {
      "type": "string",
      "description": "The name of the project's tsconfig file that includes the runtime source files. If not provided, it will default to `tsconfig.lib.json` for libraries and `tsconfig.app.json` for applications."
    }
  },
  "required": []
}
