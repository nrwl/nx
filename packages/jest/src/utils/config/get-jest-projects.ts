import { createProjectGraphAsync, type TargetConfiguration } from '@nx/devkit';
import { join, parse } from 'path';
import * as yargs from 'yargs-parser';

function getJestConfigProjectPath(projectJestConfigPath: string): string {
  return join('<rootDir>', projectJestConfigPath);
}

/**
 * Get a list of paths to all the jest config files
 * using the Nx Jest executor and `@nx/run:commands`
 * running `jest`.
 *
 * This is used to configure Jest multi-project support.
 *
 * To add a project not using the Nx Jest executor:
 * export default async () => ({
 *   projects: [...(await getJestProjectsAsync()), '<rootDir>/path/to/jest.config.ts'];
 * });
 *
 **/
export async function getJestProjectsAsync() {
  const graph = await createProjectGraphAsync({
    exitOnError: false,
    resetDaemonClient: true,
  });
  const jestConfigurations = new Set<string>();
  for (const node of Object.values(graph.nodes)) {
    const projectConfig = node.data;
    if (!projectConfig.targets) {
      continue;
    }
    for (const targetConfiguration of Object.values(projectConfig.targets)) {
      if (
        targetConfiguration.executor === '@nx/jest:jest' ||
        targetConfiguration.executor === '@nrwl/jest:jest'
      ) {
        collectJestConfigFromJestExecutor(
          targetConfiguration,
          jestConfigurations
        );
      } else if (targetConfiguration.executor === 'nx:run-commands') {
        collectJestConfigFromRunCommandsExecutor(
          targetConfiguration,
          projectConfig.root,
          jestConfigurations
        );
      }
    }
  }

  removeDuplicates(jestConfigurations);
  return Array.from(jestConfigurations);
}

// If two paths result in same project, prefer the more specific path.
// e.g. <rootDir>/demo/jest.config.js over <rootDir>/demo
function removeDuplicates(configs: Set<string>): void {
  configs.forEach((config) => {
    const { dir, ext } = parse(config);
    // If the directory has been added previously, remove it and keep the current, more specific path.
    if (ext) configs.delete(dir);
  });
}

function collectJestConfigFromJestExecutor(
  targetConfiguration: TargetConfiguration,
  jestConfigurations: Set<string>
): void {
  if (targetConfiguration.options?.jestConfig) {
    jestConfigurations.add(
      getJestConfigProjectPath(targetConfiguration.options.jestConfig)
    );
  }
  if (targetConfiguration.configurations) {
    for (const configurationObject of Object.values(
      targetConfiguration.configurations
    )) {
      if (configurationObject.jestConfig) {
        jestConfigurations.add(
          getJestConfigProjectPath(configurationObject.jestConfig)
        );
      }
    }
  }
}

function collectJestConfigFromRunCommandsExecutor(
  targetConfiguration: TargetConfiguration,
  projectRoot: string,
  jestConfigurations: Set<string>
): void {
  if (targetConfiguration.options?.command) {
    collectJestConfigFromCommand(
      targetConfiguration.options.command,
      targetConfiguration.options.cwd ?? projectRoot,
      jestConfigurations
    );
  } else if (targetConfiguration.options?.commands) {
    for (const command of targetConfiguration.options.commands) {
      const commandScript =
        typeof command === 'string' ? command : command.command;
      collectJestConfigFromCommand(
        commandScript,
        targetConfiguration.options.cwd ?? projectRoot,
        jestConfigurations
      );
    }
  }

  if (targetConfiguration.configurations) {
    for (const configurationObject of Object.values(
      targetConfiguration.configurations
    )) {
      if (configurationObject.command) {
        collectJestConfigFromCommand(
          configurationObject.command,
          configurationObject.cwd ?? projectRoot,
          jestConfigurations
        );
      } else if (configurationObject.commands) {
        for (const command of configurationObject.commands) {
          const commandScript =
            typeof command === 'string' ? command : command.command;
          collectJestConfigFromCommand(
            commandScript,
            configurationObject.cwd ?? projectRoot,
            jestConfigurations
          );
        }
      }
    }
  }
}

function collectJestConfigFromCommand(
  command: string,
  cwd: string,
  jestConfigurations: Set<string>
) {
  const jestCommandRegex =
    /(?<=^|&)(?:[^&\r\n\s]* )*jest(?: [^&\r\n\s]*)*(?=$|&)/g;
  const matches = command.match(jestCommandRegex);
  if (!matches) {
    return;
  }

  for (const match of matches) {
    const parsed = yargs(match, {
      configuration: { 'strip-dashed': true },
      string: ['config'],
    });
    if (parsed.config) {
      jestConfigurations.add(
        getJestConfigProjectPath(join(cwd, parsed.config))
      );
    } else {
      jestConfigurations.add(getJestConfigProjectPath(cwd));
    }
  }
}
