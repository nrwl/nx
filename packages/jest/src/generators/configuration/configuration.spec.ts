import 'nx/src/internal-testing-utils/mock-project-graph';

import {
  addProjectConfiguration,
  readJson,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
  writeJson,
  updateJson,
  readNxJson,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { jestConfigObject } from '../../utils/config/functions';

import configurationGenerator from './configuration';
import { JestProjectSchema } from './schema.d';

describe('jestProject', () => {
  let tree: Tree;
  let defaultOptions: Omit<JestProjectSchema, 'project'> = {
    supportTsx: false,
    skipSetupFile: false,
    skipSerializers: false,
    testEnvironment: 'jsdom',
    setupFile: 'none',
    skipFormat: false,
    compiler: 'tsc',
    addPlugin: true,
  };

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    addProjectConfiguration(tree, 'lib1', {
      root: 'libs/lib1',
      sourceRoot: 'libs/lib1/src',
      targets: {
        lint: {
          executor: '@nx/eslint:lint',
          options: {},
        },
      },
    });
    writeJson(tree, 'libs/lib1/tsconfig.json', {
      files: [],
      include: [],
      references: [],
    });
  });

  it('should generate files', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      setupFile: 'angular',
    } as JestProjectSchema);
    expect(tree.read('libs/lib1/src/test-setup.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { setupZoneTestEnv } from 'jest-preset-angular/setup-env/zone';

      setupZoneTestEnv();
      "
    `);
    expect(tree.exists('libs/lib1/jest.config.ts')).toBeTruthy();
    expect(tree.exists('libs/lib1/tsconfig.spec.json')).toBeTruthy();
    expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should add target if there are no targets', async () => {
    const pc = readProjectConfiguration(tree, 'lib1');
    delete pc.targets;
    updateProjectConfiguration(tree, 'lib1', pc);
    expect(async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        setupFile: 'angular',
      });
    }).not.toThrow();
  });

  it('should create a jest.config.ts', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
    } as JestProjectSchema);
    expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should add a reference to solution tsconfig.json', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
    } as JestProjectSchema);
    const tsConfig = readJson(tree, 'libs/lib1/tsconfig.json');
    expect(tsConfig.references).toContainEqual({
      path: './tsconfig.spec.json',
    });
  });

  it('should create a tsconfig.spec.json', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      setupFile: 'angular',
    } as JestProjectSchema);
    const tsConfig = readJson(tree, 'libs/lib1/tsconfig.spec.json');
    expect(tsConfig).toEqual({
      extends: './tsconfig.json',
      compilerOptions: {
        module: 'commonjs',
        outDir: '../../dist/out-tsc',
        types: ['jest', 'node'],
        target: 'es2016',
      },
      files: ['src/test-setup.ts'],
      include: [
        'jest.config.ts',
        'src/**/*.test.ts',
        'src/**/*.spec.ts',
        'src/**/*.d.ts',
      ],
    });
  });

  describe('--setup-file', () => {
    it('should generate src/test-setup.ts', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
      } as JestProjectSchema);
      expect(tree.exists('src/test-setup.ts')).toBeFalsy();
      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).not.toContain(
        `setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],`
      );
    });

    it('should have setupFilesAfterEnv in the jest.config when generated for web-components', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        setupFile: 'web-components',
      } as JestProjectSchema);
      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).toContain(
        `setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],`
      );
    });

    it('should have setupFilesAfterEnv and globals.ts-jest in the jest.config when generated for angular', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        setupFile: 'angular',
      } as JestProjectSchema);

      const jestConfig = tree.read('libs/lib1/jest.config.ts', 'utf-8');
      expect(jestConfig).toContain(
        `setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],`
      );
      expect(jestConfig).toMatchSnapshot();
    });

    it('should not list the setup file in tsconfig.spec.json', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        setupFile: 'none',
      } as JestProjectSchema);
      const tsConfig = readJson(tree, 'libs/lib1/tsconfig.spec.json');
      expect(tsConfig.files).toBeUndefined();
    });
  });

  describe('--skip-setup-file', () => {
    it('should generate src/test-setup.ts', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        skipSetupFile: true,
      } as JestProjectSchema);
      expect(tree.exists('src/test-setup.ts')).toBeFalsy();
    });

    it('should not list the setup file in tsconfig.spec.json', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        skipSetupFile: true,
      } as JestProjectSchema);
      const tsConfig = readJson(tree, 'libs/lib1/tsconfig.spec.json');
      expect(tsConfig.files).toBeUndefined();
    });
  });

  describe('--skip-serializers', () => {
    it('should not list the serializers in jest.config.ts', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        skipSerializers: true,
      } as JestProjectSchema);
      const jestConfig = tree.read('libs/lib1/jest.config.ts', 'utf-8');
      expect(jestConfig).not.toContain(`
      snapshotSerializers: [
        'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js,
        'jest-preset-angular/build/AngularSnapshotSerializer.js',
        'jest-preset-angular/build/HTMLCommentSerializer.js'
      ]
    `);
    });
  });

  describe('--support-tsx', () => {
    it('should add jest.transform', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        supportTsx: true,
      } as JestProjectSchema);
      const jestConfig = jestConfigObject(tree, 'libs/lib1/jest.config.ts');
      expect(jestConfig.transform).toEqual({
        '^.+\\.[tj]sx?$': [
          'ts-jest',
          { tsconfig: '<rootDir>/tsconfig.spec.json' },
        ],
      });
    });

    it('should add tsx to moduleExtensions', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        supportTsx: true,
      } as JestProjectSchema);
      const jestConfig = jestConfigObject(tree, 'libs/lib1/jest.config.ts');
      expect(jestConfig.moduleFileExtensions).toEqual([
        'ts',
        'tsx',
        'js',
        'jsx',
      ]);
    });
  });

  it('should create jest.config.js with --js flag', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      js: true,
    } as JestProjectSchema);
    expect(tree.exists('jest.preset.js')).toBeTruthy();
    expect(tree.exists('jest.config.js')).toBeTruthy();
    expect(tree.exists('libs/lib1/jest.config.js')).toBeTruthy();
    expect(tree.read('libs/lib1/jest.config.js', 'utf-8')).toContain(
      "preset: '../../jest.preset.js',"
    );
  });

  it('should use jest.config.js in project config with --js flag', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      js: true,
    } as JestProjectSchema);
    expect(tree.exists('libs/lib1/jest.config.js')).toBeTruthy();
  });

  it('should generate a jest.preset.js when it does not exist', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      js: true,
    } as JestProjectSchema);
    expect(tree.exists('libs/lib1/jest.config.js')).toBeTruthy();
    expect(tree.exists('jest.preset.js')).toBeTruthy();
    expect(tree.read('libs/lib1/jest.config.js', 'utf-8')).toContain(
      "preset: '../../jest.preset.js',"
    );
  });

  it('should not override existing jest preset file and should point to it in jest.config files', async () => {
    tree.write('jest.preset.mjs', 'export default {}');
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      js: true,
    } as JestProjectSchema);
    expect(tree.exists('libs/lib1/jest.config.js')).toBeTruthy();
    expect(tree.exists('jest.preset.mjs')).toBeTruthy();
    expect(tree.read('libs/lib1/jest.config.js', 'utf-8')).toContain(
      "preset: '../../jest.preset.mjs',"
    );
  });

  it('should use module.exports with --js flag', async () => {
    await configurationGenerator(tree, {
      ...defaultOptions,
      project: 'lib1',
      js: true,
    } as JestProjectSchema);
    expect(tree.exists('libs/lib1/jest.config.js')).toBeTruthy();
    expect(tree.read('libs/lib1/jest.config.js', 'utf-8')).toContain(
      'module.exports = {'
    );
  });

  describe('--babelJest', () => {
    it('should generate proper jest.transform when babelJest is true', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        babelJest: true,
      } as JestProjectSchema);
      const jestConfig = jestConfigObject(tree, 'libs/lib1/jest.config.ts');

      expect(jestConfig.globals).not.toEqual({
        'ts-jest': {
          tsConfig: '<rootDir>/tsconfig.spec.json',
        },
      });
      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).toMatchSnapshot();
    });

    it('should generate proper jest.transform when babelJest and supportTsx is true', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        babelJest: true,
        supportTsx: true,
      } as JestProjectSchema);
      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).toMatchSnapshot();
    });

    it('should generate proper jest.transform when --compiler=swc and supportTsx is true', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
        compiler: 'swc',
        supportTsx: true,
      } as JestProjectSchema);

      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8')).toMatchSnapshot();
      // assert the TS solution setup doesn't leak into the old/integrated setup
      expect(tree.exists('libs/lib1/.spec.swcrc')).toBeFalsy();
    });
  });

  describe('root project', () => {
    it('root jest.config.ts should be project config', async () => {
      writeJson(tree, 'tsconfig.json', {
        files: [],
        include: [],
        references: [],
      });
      addProjectConfiguration(tree, 'my-project', {
        root: '',
        sourceRoot: 'src',
        name: 'my-project',
        targets: {},
      });
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'my-project',
      });
      expect(tree.read('jest.config.ts', 'utf-8')).toMatchInlineSnapshot(`
        "export default {
          displayName: 'my-project',
          preset: './jest.preset.js',
          coverageDirectory: './coverage/my-project',
          testMatch: [
            '<rootDir>/src/**/__tests__/**/*.[jt]s?(x)',
            '<rootDir>/src/**/*(*.)@(spec|test).[jt]s?(x)',
          ],
        };
        "
      `);
    });

    it('root jest.config.js should be project config', async () => {
      writeJson(tree, 'tsconfig.json', {
        files: [],
        include: [],
        references: [],
      });
      addProjectConfiguration(tree, 'my-project', {
        root: '',
        sourceRoot: 'src',
        name: 'my-project',
        targets: {},
      });
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'my-project',
        js: true,
      });
      expect(tree.read('jest.config.js', 'utf-8')).toMatchInlineSnapshot(`
        "module.exports = {
          displayName: 'my-project',
          preset: './jest.preset.js',
          coverageDirectory: './coverage/my-project',
          testMatch: [
            '<rootDir>/src/**/__tests__/**/*.[jt]s?(x)',
            '<rootDir>/src/**/*(*.)@(spec|test).[jt]s?(x)',
          ],
        };
        "
      `);
    });
  });

  describe(`jest.preset.cjs`, () => {
    it(`root jest.preset.cjs existing should force subsequent configs to point to it correctly`, async () => {
      // ARRANGE
      tree.write(
        `jest.preset.cjs`,
        `
      const nxPreset = require('@nx/jest/preset').default;

      module.exports = { ...nxPreset }`
      );

      // ACT
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
      } as JestProjectSchema);

      // ASSERT
      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8'))
        .toMatchInlineSnapshot(`
        "export default {
          displayName: 'lib1',
          preset: '../../jest.preset.cjs',
          coverageDirectory: '../../coverage/libs/lib1',
        };
        "
      `);
    });

    it(`root package.json type=module should create jest.preset.cjs and force subsequent configs to point to it correctly`, async () => {
      // ARRANGE
      updateJson(tree, 'package.json', (pkgJson) => {
        pkgJson.type = 'module';
        return pkgJson;
      });

      // ACT
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'lib1',
      } as JestProjectSchema);

      // ASSERT
      expect(tree.exists('jest.preset.cjs')).toBeTruthy();
      expect(tree.read('libs/lib1/jest.config.ts', 'utf-8'))
        .toMatchInlineSnapshot(`
        "export default {
          displayName: 'lib1',
          preset: '../../jest.preset.cjs',
          coverageDirectory: '../../coverage/libs/lib1',
        };
        "
      `);
    });
  });

  describe('TS solution setup', () => {
    beforeEach(() => {
      tree = createTreeWithEmptyWorkspace();
      updateJson(tree, 'package.json', (json) => {
        json.workspaces = ['packages/*'];
        return json;
      });
      writeJson(tree, 'tsconfig.base.json', {
        compilerOptions: { composite: true },
      });
      writeJson(tree, 'tsconfig.json', {
        extends: './tsconfig.base.json',
        files: [],
        references: [],
      });

      addProjectConfiguration(tree, 'pkg1', {
        root: 'packages/pkg1',
        sourceRoot: 'packages/pkg1/src',
        targets: {
          lint: {
            executor: '@nx/eslint:lint',
            options: {},
          },
        },
      });
      writeJson(tree, 'packages/pkg1/tsconfig.json', {
        files: [],
        include: [],
        references: [],
      });
    });

    it('should generate files', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'pkg1',
      });

      expect(tree.exists('packages/pkg1/tsconfig.spec.json')).toBeTruthy();
      expect(tree.exists('packages/pkg1/jest.config.ts')).toBeTruthy();
      expect(tree.read('packages/pkg1/jest.config.ts', 'utf-8'))
        .toMatchInlineSnapshot(`
        "export default {
          displayName: 'pkg1',
          preset: '../../jest.preset.js',
          coverageDirectory: 'test-output/jest/coverage',
        };
        "
      `);
      expect(tree.exists('packages/pkg1/.spec.swcrc')).toBeFalsy();
    });

    it(`should setup a task pipeline for the test target to depend on the deps' build target`, async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'pkg1',
      });

      const nxJson = readNxJson(tree);
      expect(nxJson.targetDefaults.test.dependsOn).toStrictEqual(['^build']);
    });

    it('should generate files with swc compiler', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'pkg1',
        compiler: 'swc',
      });

      expect(tree.exists('packages/pkg1/tsconfig.spec.json')).toBeTruthy();
      expect(tree.exists('packages/pkg1/jest.config.ts')).toBeTruthy();
      expect(tree.read('packages/pkg1/jest.config.ts', 'utf-8'))
        .toMatchInlineSnapshot(`
        "/* eslint-disable */
        import { readFileSync } from 'fs';

        // Reading the SWC compilation config for the spec files
        const swcJestConfig = JSON.parse(
          readFileSync(\`\${__dirname}/.spec.swcrc\`, 'utf-8')
        );

        // Disable .swcrc look-up by SWC core because we're passing in swcJestConfig ourselves
        swcJestConfig.swcrc = false;

        export default {
          displayName: 'pkg1',
          preset: '../../jest.preset.js',
          transform: {
            '^.+\\\\.[tj]s$': ['@swc/jest', swcJestConfig],
          },
          moduleFileExtensions: ['ts', 'js', 'html'],
          coverageDirectory: 'test-output/jest/coverage',
        };
        "
      `);
      expect(tree.exists('packages/pkg1/.spec.swcrc')).toBeTruthy();
      expect(tree.read('packages/pkg1/.spec.swcrc', 'utf-8'))
        .toMatchInlineSnapshot(`
        "{
          "jsc": {
            "target": "es2017",
            "parser": {
              "syntax": "typescript",
              "decorators": true,
              "dynamicImport": true
            },
            "transform": {
              "decoratorMetadata": true,
              "legacyDecorator": true
            },
            "keepClassNames": true,
            "externalHelpers": true,
            "loose": true
          },
          "module": {
            "type": "es6"
          },
          "sourceMaps": true,
          "exclude": []
        }
        "
      `);
    });

    it('should generate the correct options for swc when "supportTsx: true"', async () => {
      await configurationGenerator(tree, {
        ...defaultOptions,
        project: 'pkg1',
        compiler: 'swc',
        supportTsx: true,
      });

      expect(tree.read('packages/pkg1/.spec.swcrc', 'utf-8'))
        .toMatchInlineSnapshot(`
        "{
          "jsc": {
            "target": "es2017",
            "parser": {
              "syntax": "typescript",
              "decorators": true,
              "dynamicImport": true,
              "tsx": true
            },
            "transform": {
              "decoratorMetadata": true,
              "legacyDecorator": true,
              "react": {
                "runtime": "automatic"
              }
            },
            "keepClassNames": true,
            "externalHelpers": true,
            "loose": true
          },
          "module": {
            "type": "es6"
          },
          "sourceMaps": true,
          "exclude": []
        }
        "
      `);
    });
  });
});
