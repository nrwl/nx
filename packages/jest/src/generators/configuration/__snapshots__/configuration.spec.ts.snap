// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jestProject --babelJest should generate proper jest.transform when --compiler=swc and supportTsx is true 1`] = `
"export default {
  displayName: 'lib1',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\\\.[tj]sx?$': [
      '@swc/jest',
      {
        jsc: {
          parser: { syntax: 'typescript', tsx: true },
          transform: { react: { runtime: 'automatic' } },
        },
      },
    ],
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib1',
};
"
`;

exports[`jestProject --babelJest should generate proper jest.transform when babelJest and supportTsx is true 1`] = `
"export default {
  displayName: 'lib1',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib1',
};
"
`;

exports[`jestProject --babelJest should generate proper jest.transform when babelJest is true 1`] = `
"export default {
  displayName: 'lib1',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\\\.[tj]s$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/libs/lib1',
};
"
`;

exports[`jestProject --setup-file should have setupFilesAfterEnv and globals.ts-jest in the jest.config when generated for angular 1`] = `
"export default {
  displayName: 'lib1',
  preset: '../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  coverageDirectory: '../../coverage/libs/lib1',
  transform: {
    '^.+\\\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!.*\\\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
"
`;

exports[`jestProject should create a jest.config.ts 1`] = `
"export default {
  displayName: 'lib1',
  preset: '../../jest.preset.js',
  coverageDirectory: '../../coverage/libs/lib1',
};
"
`;

exports[`jestProject should generate files 2`] = `
"export default {
  displayName: 'lib1',
  preset: '../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  coverageDirectory: '../../coverage/libs/lib1',
  transform: {
    '^.+\\\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!.*\\\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
"
`;
